src/transform/src/analysis.rs

```diff
@@ -493,6 +493,97 @@ pub mod common {
+            results: &[Self::Value],
+            depends: &Derived,
+        ) -> Self::Value {
+            let mut child_results = depends
+                .children_of_rev(index, expr.children().count())
+                .map(|child| results[child])
+                .collect::<Vec<_>>();
+            child_results.reverse();
+            match expr {
+                MirRelationExpr::Get { id, .. } => match id {
+                    Id::Global(id) => self.global_mon_ids.contains(id),
+                    Id::Local(id) => *results
+                        .get(
+                            *depends
+                                .bindings()
+                                .get(id)
+                                .expect("Dependency info not found"),
+                        )
+                        .unwrap_or(&false),
+                },
+                MirRelationExpr::Project { .. } => child_results[0],
+                MirRelationExpr::Filter { predicates, .. } => {
+                    let is_monotonic = child_results[0];
+                    // Temporal predicates can introduce non-monotonicity, as they
+                    // can result in the future removal of records.
+                    // TODO: this could be improved to only restrict if upper bounds
+                    // are present, as temporal lower bounds only delay introduction.
+                    is_monotonic && !predicates.iter().any(|p| p.contains_temporal())
+                }
+                MirRelationExpr::Map { .. } => child_results[0],
+                MirRelationExpr::TopK { .. } => false,
+                MirRelationExpr::Reduce { aggregates, .. } => {
+                    // Reduce is monotonic iff its input is and it is a "distinct",
+                    // with no aggregate values; otherwise it may need to retract.
+                    child_results[0] && aggregates.is_empty()
+                }
+                MirRelationExpr::Union { .. } => {
+                    child_results.iter().fold(true, |acc, item| acc && *item)
+                }
+                MirRelationExpr::ArrangeBy { .. } => child_results[0],
+                MirRelationExpr::FlatMap { func, .. } => {
+                    child_results[0] && func.preserves_monotonicity()
+                }
+                MirRelationExpr::Join { .. } => {
+                    // If all inputs to the join are monotonic then so is the join.
+                    child_results.iter().fold(true, |acc, item| acc && *item)
+                }
+                MirRelationExpr::Threshold { .. } => child_results[0],
+                MirRelationExpr::Let { .. } | MirRelationExpr::LetRec { .. } => results[index - 1],
+                MirRelationExpr::Negate { .. } => false,
+                MirRelationExpr::Constant { rows: Ok(rows), .. } => {
+                    rows.iter().all(|(_, diff)| diff > &0)
+                }
+                MirRelationExpr::Constant { rows: Err(_), .. } => false,
```

Comment:
Can we document this case? I'm not sure why it is `false`.

Reply:
We just took this from the old code without thinking too much, but I have to admit I also don't see a reason for this to be `false`.

Reply:
Opened an issue about this for now: #28499